src/client/csfle/client_encryption.rs

```diff
@@ -0,0 +1,294 @@
+        let crypt = Crypt::builder().build()?;
+        let exec = CryptExecutor::new(
+            opts.key_vault_client.weak(),
+            opts.key_vault_namespace.clone(),
+            None,
+            None,
+            opts.tls_options,
+        )?;
+        let key_vault = opts
+            .key_vault_client
+            .database(&opts.key_vault_namespace.db)
+            .collection_with_options(
+                &opts.key_vault_namespace.coll,
+                CollectionOptions::builder()
+                    .write_concern(WriteConcern::MAJORITY)
+                    .read_concern(ReadConcern::MAJORITY)
+                    .build(),
+            );
+        Ok(Self {
+            crypt,
+            exec,
+            key_vault,
+        })
+    }
+
+    /// Creates a new key document and inserts into the key vault collection.
+    /// Returns the _id of the created document as a UUID (BSON binary subtype 0x04).
+    pub async fn create_data_key(
+        &self,
+        kms_provider: KmsProvider,
+        opts: DataKeyOptions,
+    ) -> Result<Binary> {
+        let ctx = self.create_data_key_ctx(kms_provider, opts)?;
+        let data_key = self.exec.run_ctx(ctx, None).await?;
+        self.key_vault.insert_one(&data_key, None).await?;
+        let bin_ref = data_key
+            .get_binary("_id")
+            .map_err(|e| Error::internal(format!("invalid data key id: {}", e)))?;
+        Ok(bin_owned(bin_ref))
+    }
+
+    fn create_data_key_ctx(&self, kms_provider: KmsProvider, opts: DataKeyOptions) -> Result<Ctx> {
+        let mut builder = self.crypt.ctx_builder();
+        let mut key_doc = doc! { "provider": kms_provider.name() };
+        if let Some(master_key) = opts.master_key {
+            key_doc.extend(master_key);
+        }
+        builder = builder.key_encryption_key(&key_doc)?;
+        if let Some(alt_names) = opts.key_alt_names {
+            for name in alt_names {
+                builder = builder.key_alt_name(&name)?;
+            }
+        }
+        if let Some(material) = opts.key_material {
+            builder = builder.key_material(&material)?;
+        }
+        Ok(builder.build_datakey()?)
+    }
+
+    // pub async fn rewrap_many_data_key(&self, _filter: Document, _opts: RewrapManyDataKeyOptions)
+    // -> Result<RewrapManyDataKeyResult> { todo!("RUST-1441")
+    // }
+
+    /// Removes the key document with the given UUID (BSON binary subtype 0x04) from the key vault
+    /// collection. Returns the result of the internal deleteOne() operation on the key vault
+    /// collection.
+    pub async fn delete_key(&self, id: &Binary) -> Result<DeleteResult> {
+        self.key_vault.delete_one(doc! { "_id": id }, None).await
+    }
+
+    /// Finds a single key document with the given UUID (BSON binary subtype 0x04).
+    /// Returns the result of the internal find() operation on the key vault collection.
+    pub async fn get_key(&self, id: &Binary) -> Result<Option<RawDocumentBuf>> {
```

Comment:
This API uses raw types because it seemed likely that a user performing encryption/decryption would want to avoid the additional overhead of parsing.  This was a handwavy judgement call that I'd be happy to revisit if people feel strongly the other way.